allium/lib/diversity_index.py

```
@@ -0,0 +1,222 @@
+def _log_ramp(value, cap):
+    """Log-saturating ramp: log(1+value)/log(1+cap), clamped to [0, 1].
+
+    Rewards early growth strongly and large fleets with diminishing returns,
+    so raw size cannot drown the distinct-count halves.
+    """
+    if cap <= 0:
+        return 0.0
+    return min(math.log(1 + max(value, 0)) / math.log(1 + cap), 1.0)
+
+
+def _rank_n_value(values, n=VOLUME_CAP_RANK, floor=VOLUME_CAP_FLOOR):
```

Ruff (0.15.20)

[warning] Missing return type annotation for private function `_rank_n_value`

(ANN202)